applications/MeshingApplication/custom_utilities/uniform_refine_utility.cpp

```diff
@@ -0,0 +1,680 @@
+
+    mSubModelPartsColors.ComputeSubModelPartsList(mNodesColorMap, mCondColorMap, mElemColorMap, mColors);
+}
+
+
+/// Destructor
+template< unsigned int TDim>
+UniformRefineUtility<TDim>::~UniformRefineUtility() {}
+
+
+/// Turn back information as a string.
+template< unsigned int TDim>
+std::string UniformRefineUtility<TDim>::Info() const {
+    return "Uniform refine utility.";
+}
+
+
+/// Print information about this object.
+template< unsigned int TDim>
+void UniformRefineUtility<TDim>::PrintInfo(std::ostream& rOStream) const {
+    rOStream << "Uniform refine utility.";
+}
+
+
+/// Print object's data.
+template< unsigned int TDim>
+void UniformRefineUtility<TDim>::PrintData(std::ostream& rOStream) const {
+    rOStream << "Uniform refine utility constructed with:\n";
+    rOStream << "   Model part: " << mrModelPart.Info() << "\n";
+    rOStream << "   Final refinement level: " << mFinalRefinementLevel << "\n";
+}
+
+
+/// Execute the refinement until the final refinement level is reached
+template< unsigned int TDim>
+void UniformRefineUtility<TDim>::Refine()
+{
+    // Get the lowest refinement level
+    int minimum_refinement_level = 1e6;
+    const int n_elements = mrModelPart.Elements().size();
+    for (int i = 0; i < n_elements; i++)
+    {
+        ModelPart::ElementsContainerType::iterator ielement = mrModelPart.ElementsBegin() + i;
+        if (ielement->GetValue(REFINEMENT_LEVEL) < minimum_refinement_level)
+            minimum_refinement_level = ielement->GetValue(REFINEMENT_LEVEL);
+    }
+
+    for (int level = minimum_refinement_level; level < mFinalRefinementLevel; level++)
+    {
+        RefineLevel(level);
+    }
+}
+
+
+/// Execute the refinement once
+template< unsigned int TDim>
+void UniformRefineUtility<TDim>::RefineLevel(const int& rThisLevel)
+{
+    // Initialize the entities Id lists
+    std::vector<int> elements_id;
+    std::vector<int> conditions_id;
+
+
+    // Get the elements id
+    const int n_elements = mrModelPart.Elements().size();
+    for (int i = 0; i < n_elements; i++)
+    {
+        ModelPart::ElementsContainerType::iterator ielement = mrModelPart.ElementsBegin() + i;
+
+        // Check the refinement level of the origin elements
+        int step_refine_level = ielement->GetValue(REFINEMENT_LEVEL);
+        if (step_refine_level == rThisLevel)
+            elements_id.push_back(ielement->Id());
+    }
+
+    // Get the conditions id
+    const int n_conditions = mrModelPart.Conditions().size();
+    for (int i = 0; i < n_conditions; i++)
+    {
+        ModelPart::ConditionsContainerType::iterator icondition = mrModelPart.ConditionsBegin() + i;
+
+        // Check the refinement level of the origin conditions
+        int step_refine_level = icondition->GetValue(REFINEMENT_LEVEL);
+        if (step_refine_level == rThisLevel)
+            conditions_id.push_back(icondition->Id());
+    }
+
+    // Loop the origin elements to create the middle nodes
+    for (int id : elements_id)
```

Comment:
In combination with the previous comments, if you use auto you will avoid the problem with the type in type in the iterations of containers